code/src/hooks/use-infinite-products-search/useInfiniteProductsSearch.ts

```diff
@@ -0,0 +1,80 @@
+import { skipToken } from "@reduxjs/toolkit/query/react";
+import { useCallback, useEffect, useRef, useState } from "react";
+
+import { useLocaleContext } from "@/context/i18n/I18nProvider";
+import useDebouncedValue from "@/hooks/use-debounced-value/useDebouncedValue";
+import { useGetUserProductsBySearchQuery } from "@/store/api/productsApi";
+import { GetUserProductsBySearchQueryResponse } from "@/types/product.types";
+
+type useInfiniteProductsSearchConfig = {
+  query: string;
+  pageSize?: number;
+  minQueryLength?: number;
+  debounceDelayMs?: number;
+};
+
+const useInfiniteProductsSearch = ({
+  query = "",
+  pageSize = 8,
+  minQueryLength = 4,
+  debounceDelayMs = 500
+}: useInfiniteProductsSearchConfig) => {
+  const [currentPage, setCurrentPage] = useState(0);
+  const [searchProducts, setSearchProducts] =
+    useState<GetUserProductsBySearchQueryResponse>();
+  const isLastPageRef = useRef(false);
+
+  const { locale } = useLocaleContext();
+
+  const debouncedQuery = useDebouncedValue(query, debounceDelayMs);
+
+  const queryParameters =
+    debouncedQuery.length >= minQueryLength
+      ? {
+          searchQuery: debouncedQuery,
+          lang: locale,
+          page: currentPage,
+          size: pageSize
+        }
+      : skipToken;
+
+  const { data, isLoading, isError } =
+    useGetUserProductsBySearchQuery(queryParameters);
+
+  const resetSearch = useCallback(() => {
+    setCurrentPage(0);
+    setSearchProducts(undefined);
+  }, []);
+
+  useEffect(() => {
+    resetSearch();
+  }, [debouncedQuery]);
+
+  useEffect(() => {
+    if (data?.content) {
+      setSearchProducts((prev) => {
+        const wasSearchReset = prev !== undefined;
+
+        return wasSearchReset
+          ? { ...data, content: [...prev.content, ...data.content] }
+          : data;
+      });
```

Comment:
in your case you laready check if `wasSearchReset` true
this check `const wasSearchReset = prev !== undefined` is useless

can write it simpler

```suggestion
setSearchProducts((prev) => prev
        ? { ...data, content: [...prev.content, ...data.content] }
        : data)
      isLastPageRef.current = data.last
```

Reply:
I did it for the better readability. This hook became very complicated so I tried to increase readability for better support and reusing it, also I did it at the suggestion of our teammates.

Reply:
> I did it for the better readability. This hook became very complicated so I tried to increase readability for better support and reusing it, also I did it at the suggestion of our teammates.

Don't heard. but ok. in my case it will be easier